base/cvd/cuttlefish/host/commands/cvdalloc/cvdalloc.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string.h>
+#include <unistd.h>
+
+#include <android-base/logging.h>
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/strings/str_format.h"
+
+#include "allocd/alloc_utils.h"
+#include "cuttlefish/common/libs/fs/shared_fd.h"
+
+ABSL_FLAG(int, id, 0, "Id");
+ABSL_FLAG(int, teardown_event_socket, 0, "Socket");
+ABSL_FLAG(int, allocate_event_socket, 0, "Socket");
+
+static void usage() {
+  LOG(ERROR) << "cvdalloc --id=id --allocate_event_socket=fd1 "
+                "--teardown_event_socket=fd2";
+  LOG(ERROR) << "Should only be invoked from run_cvd.";
+}
+
+namespace cuttlefish {
+
+Result<int> CvdallocMain(int argc, char *argv[]) {
+  std::vector<char *> args = absl::ParseCommandLine(argc, argv);
+
+  if (absl::GetFlag(FLAGS_id) == 0 ||
+      absl::GetFlag(FLAGS_teardown_event_socket) == 0 ||
+      absl::GetFlag(FLAGS_allocate_event_socket) == 0) {
+    usage();
+    return 1;
+  }
+
+  /*
+   * Explicit setuid calls seem to be required.
+   *
+   * This likely has something to do with invoking external commands,
+   * but it isn't clear why an explicit setuid(0) is necessary.
+   * It's possible a Linux kernel bug around permissions checking on tap
+   * devices may be the culprit, which we can't control.
+   *
+   * TODO: We should just use capabilities on Linux and rely on setuid as
+   * a fallback for other platforms.
+   *
+   * Save our current uid, so we can restore it to drop privileges later.
+   */
+  uid_t orig = getuid();
+
+  int id = absl::GetFlag(FLAGS_id);
+
+  pid_t pid = fork();
+  if (pid < 0) return -1;
+
+  if (pid > 0) {
+    int r = setuid(0);
+    if (r == -1) {
+      LOG(ERROR) << "Couldn't setuid root: " << strerror(errno);
+      return 1;
+    }
+
+    LOG(INFO) << "cvdalloc: allocating network resources";
+
+    CreateBridge("cvd-br");
+    CreateMobileIface(absl::StrFormat("cvd-mtap-%02d", id), id, kMobileIp);
```

Comment:
This is probably work for another PR as it potentially involves changing allocd's interface.

These functions should return a `Result` as creating a network interface could theoretically fail. Handling of those failures should involve tearing down the interfaces already created. I think the best way to do this, while addressing my other comment about code duplication, is to have these functions return `Result<SomeInterfaceObject>` where `SomeInterfaceObject` uses RAII to deallocate the underlying resources when its destructor runs.

Reply:
They do return `bool`s already, maybe worth at least logging failures here or some kind of fail-fast logic even without full cleanup.

Reply:
For now, I've peppered in `CF_EXPECT`s where necessary and `Teardown` if any of them fail.